Sources/MapboxDirections/VisualInstructionComponent.swift

```diff
@@ -0,0 +1,254 @@
+         - parameter format: The file format of the image. If this argument is unspecified, PNG is used.
+         - returns: A remote URL to the image.
+         */
+        public func imageURL(scale: CGFloat? = nil, format: Format = .png) -> URL? {
+            guard let imageBaseURL = imageBaseURL,
+                var imageURLComponents = URLComponents(url: imageBaseURL, resolvingAgainstBaseURL: false) else {
+                return nil
+            }
+            imageURLComponents.path += "@\(Int(scale ?? ImageRepresentation.currentScale))x.png"
+            return imageURLComponents.url
+        }
+        
+        /**
+         Returns the current screen’s native scale factor.
+         */
+        static var currentScale: CGFloat {
+            let scale: CGFloat
+            #if os(macOS)
+            scale = NSScreen.main?.backingScaleFactor ?? 1
+            #elseif os(watchOS)
+            scale = WKInterfaceDevice.current().screenScale
+            #else
+            scale = UIScreen.main.scale
+            #endif
+            return scale
+        }
+    }
+}
+
+extension VisualInstruction.Component: Codable {
+    private enum CodingKeys: String, CodingKey {
+        case kind = "type"
+        case text
+        case abbreviatedText = "abbr"
+        case abbreviatedTextPriority = "abbr_priority"
+        case imageBaseURL
+        case directions
+        case isActive = "active"
+    }
+    
+    enum Kind: String, Codable {
+        case delimiter
+        case text
+        case image = "icon"
+        case exit
+        case exitCode = "exit-number"
+        case lane
+    }
+    
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        let kind = try container.decode(Kind.self, forKey: .kind)
+        
+        if kind == .lane {
+            let indications = try container.decode(LaneIndication.self, forKey: .directions)
+            let isUsable = try container.decode(Bool.self, forKey: .isActive)
+            self = .lane(indications: indications, isUsable: isUsable)
+            return
+        }
+        
+        let text = try container.decode(String.self, forKey: .text)
+        let abbreviation = try container.decodeIfPresent(String.self, forKey: .abbreviatedText)
+        let abbreviationPriority = try container.decodeIfPresent(Int.self, forKey: .abbreviatedTextPriority)
+        let textRepresentation = TextRepresentation(text: text, abbreviation: abbreviation, abbreviationPriority: abbreviationPriority)
+        
+        switch kind {
+        case .delimiter:
+            self = .delimiter(text: textRepresentation)
+        case .text:
+            self = .text(text: textRepresentation)
+        case .image:
+            var imageBaseURL: URL?
+            if let imageBaseURLString = try container.decodeIfPresent(String.self, forKey: .imageBaseURL) {
+                imageBaseURL = URL(string: imageBaseURLString)
```

Comment:
This test fails because `imageBaseURLString` goes out of scope, somehow making `imageBaseURL` nil too.

```
Test Case '-[MapboxDirectionsTests.VisualInstructionComponentTests testImageComponent]' started.
/Users/distiller/project/Tests/MapboxDirectionsTests/VisualInstructionComponentTests.swift:39: error: -[MapboxDirectionsTests.VisualInstructionComponentTests testImageComponent] : XCTAssertEqual failed: ("Optional("https://s3.amazonaws.com/mapbox/shields/v3/us-42@1x.png")") is not equal to ("Optional("https://s3.amazonaws.com/mapbox/shields/v3/us-42@1x.svg")")
/Users/distiller/project/Tests/MapboxDirectionsTests/VisualInstructionComponentTests.swift:40: error: -[MapboxDirectionsTests.VisualInstructionComponentTests testImageComponent] : XCTAssertEqual failed: ("Optional("https://s3.amazonaws.com/mapbox/shields/v3/us-42@3x.png")") is not equal to ("Optional("https://s3.amazonaws.com/mapbox/shields/v3/us-42@3x.svg")")
Test Case '-[MapboxDirectionsTests.VisualInstructionComponentTests testImageComponent]' failed (0.010 seconds).
```

I think it may be an optimization bug in the compiler, because if I change the code to this instead, the debugger is unable to resolve `imageBaseURL` within the `if` statement:

```swift
            var imageRepresentation: ImageRepresentation
            if let imageBaseURLString = try container.decodeIfPresent(String.self, forKey: .imageBaseURL) {
                let imageBaseURL = URL(string: imageBaseURLString)
                imageRepresentation = ImageRepresentation(imageBaseURL: imageBaseURL)
            } else {
                imageRepresentation = ImageRepresentation(imageBaseURL: nil)
            }
```

Reply:
I got around that by decomposing the code a bit more, but something about this String is still tripping up `URL(string:)`, causing it to create a URL structure that itself is `nil` even though all its members are present. Meanwhile, **NS**URL is copacetic:

```
(lldb) p imageBaseURLString
(String) $R22 = "https://s3.amazonaws.com/mapbox/shields/v3/us-42"
(lldb) p URL(string: imageBaseURLString)
(URL?) $R24 = nil
(lldb) p URL(string: imageBaseURLString)?.absoluteString
(String?) $R26 = "https://s3.amazonaws.com/mapbox/shields/v3/us-42"
(lldb) po NSURL(string: imageBaseURLString)
▿ Optional<NSURL>
  - some : https://s3.amazonaws.com/mapbox/shields/v3/us-42
```

Reply:
I can easily reproduce the issue with similar test code anywhere in the example application:

```swift
        let componentJSON = [
            "text": "US 42",
            "type": "icon",
            "imageBaseURL": "https://s3.amazonaws.com/mapbox/shields/v3/us-42",
        ]
        let componentData = try! JSONSerialization.data(withJSONObject: componentJSON, options: [])
        let component = try! JSONDecoder().decode(VisualInstruction.Component.self, from: componentData)
        switch component {
        case .image(let image, _):
            image.imageBaseURL
            image.imageBaseURL?.absoluteString
            image.imageURL(scale: 1, format: .svg)
            image.imageURL(scale: 1, format: .svg)?.absoluteString
        default:
            preconditionFailure()
        }
```

The issue also reproduces if I copy the entire VisualInstruction type hierarchy into the example application as MyVisualInstruction and run the same code. However, I’m unable to reproduce the issue by copying the same code into a playground independent of MapboxDirections.swift.

Reply:
@fabian-guerra pointed out that the actual failure was a file extension mismatch rather than the URL being `nil`, which led to the fix in 53c71654c96c1b47f6910905391c220bb7ff40e7.

What’s still weird is that, when stopped inside `VisualInstruction.Component.init(decoder:)`, the debugger always prints `URL?` instances as `nil` and `URL` instances as `{}`, even if they actually have `absoluteString` set. I’ve been unable to reproduce this issue outside of that specific context, however.